### ISSUE TITLE
Web build performance: Tracking issue

### ISSUE THREAD

**Issue by valerybugakov** [sourcegraph/sourcegraph#21853]
## Description

To improve the front-end developer experience, we'd like to focus on improving build times for Storybook and the main web app bundle. 

### Legend

- 👩 Customer issue
- 🐛 Bug
- 🧶 Technical debt
- 🎩 Quality of life
- 🛠️ [Roadmap](https://docs.google.com/document/d/1cBsE9801DcBF9chZyMnxRdolqM_1c2pPyGQz15QAvYI/edit#heading=h.5nwl5fv52ess)
- 🕵️ [Spike](https://en.wikipedia.org/wiki/Spike_(software_development))
- 🔒 Security issue
- :shipit: Pull Request

**Comment by umpox** [comment]
Adding some general thoughts:

- Thought: It'd be great to be able to use esbuild and I think we'd see massive performance gains, but we might have a reliance on certain Webpack plugins that will not be supported through esbuild
- Note: We have a [branch](https://github.com/sourcegraph/sourcegraph/compare/esbuild-loader) that integrates [esbuild-loader](https://github.com/privatenumber/esbuild-loader) for use instead of `terser` and `babel`. Findings: faster build perf, slightly bigger bundle size, no hot reloading (not easily supported afaik). Still could use Webpack plugins
- Thought: I know there have been a few discussions over using native ES modules through [Vite](https://vitejs.dev/guide/), I believe this would require us to migrate to using Rollup for our builds. If we decide to go down this route, it might be worth investigating using [Snowpack](https://www.snowpack.dev/) which I believe should still support our existing Webpack config (and plugins)

**Comment by valerybugakov** [comment]
## The current state

For future reference based on [the Slack discussion](https://sourcegraph.slack.com/archives/C07KZF47K/p1625522218184200).

### Did we try/consider esbuild?

Yes, @sourcegraph/frontend-platform considered Esbuild and <a target="_blank" href="https://github.com/sourcegraph/sourcegraph/compare/esbuild-loader" rel="noopener noreferrer">tried</a> using <a target="_blank" href="https://github.com/privatenumber/esbuild-loader" rel="noopener noreferrer">esbuild-loader</a> in our Webpack config. 

### Issues we see

Esbuild doesn't support **hot-module-replacement** yet. This is one of the main blockers. Esbuild performance is awesome, but it's important for DX to bypass page load time and API requests when working on the web app UI. Ryan Florence left <a target="_blank" href="https://github.com/evanw/esbuild/issues/464#issuecomment-794626479" rel="noopener noreferrer">a great comment</a> about it. It seems that Vite <a target="_blank" href="https://vitejs.dev/guide/features.html#hot-module-replacement" rel="noopener noreferrer">got the best from both worlds</a> but still has many small inconveniences compared to Webpack (based on the GH issues and hacker news discussions), which are the result of an immature plugins ecosystem.

Another important consideration: apart from the web app, we have Storybook and browser extension powered by Webpack. We want to **keep one tool for all our web applications** because it results in a predictable behavior, and it's easier to support one instrument. Storybook just <a target="_blank" href="https://storybook.js.org/blog/storybook-for-webpack-5/#next-gen-bundlers" rel="noopener noreferrer">started</a> working on the experimental API for custom builders like <a target="_blank" href="https://github.com/eirslett/storybook-builder-vite" rel="noopener noreferrer">Vite</a> but doesn't support raw Esbuild yet.

Last but not least, complete migration requires quite a bit of **time**. We didn't have a chance to explore all the available options yet properly. Even Webpack 5 upgrade got stuck twice because of various issues. As @sourcegraph/frontend-platform, we allocated time in Q3 to dig into available build tools and investigate tradeoffs properly.

### Webpack improvements overview

Webpack 5 brings <a target="_blank" href="https://webpack.js.org/blog/2020-10-10-webpack-5-release/#persistent-caching" rel="noopener noreferrer">persistent caching</a>, which significantly improves build performance and recompilation. With additional tweaks like <a target="_blank" href="https://webpack.js.org/plugins/dll-plugin/#root" rel="noopener noreferrer">DLL Plugin</a>, it can be improved further. Also, we can consider <a target="_blank" href="https://webpack.js.org/concepts/module-federation/#root" rel="noopener noreferrer">module federation</a> to avoid building the whole web application in development but split it up into multiple sub-applications via Webpack 5 plugins.

Keeping Webpack (at least for now) seems like a more stable/simple way of improving our web build performance. We're definitely keeping an eye on Esbuild and other modern build tools available.

**Comment by sqs** [comment]
@valerybugakov Thank you! This is very helpful to understand. I personally found the slow recompilation so bad that I prefer fast esbuild to anything else, even if that means lack of HMR, consistency (w/storybook), etc. Maybe it's just worse on my machine, though? It takes ~5-10s to recompile and reload, and I'm on an AMD Epyc 64-core workstation with 128GB RAM.

Also, the benefits of speeding up our frontend dev seem so large that it makes sense to prioritize them really highly above other frontend work. If you agree but weren't able to convince others of the importance of this kind of work, I'd be curious to hear that.

Anyway, with all this said, I trust you to make the right call here, I just want to give my input as a drive-by dev (not as CEO).

**Comment by valerybugakov** [comment]
@sqs, definitely, we see the issue with recompilation too. Webpack 5 [upgrade](https://github.com/sourcegraph/sourcegraph/pull/22580) improves JS recompilation and [multiple other things](https://sourcegraph.slack.com/archives/C89KCDK5J/p1626765923282000). It's in the `main` branch now. Check it out!  Recompilation time will be improved further by the [DLL plugin integration](https://github.com/sourcegraph/sourcegraph/issues/22261) into the web app. 

SCSS recompilation is slow too. Right now, ~70% of stylesheets are combined into one giant file, which means that changing one rule results in bundling it again from scratch. This aspect will be addressed by [completely migrating to CSS modules](https://github.com/sourcegraph/sourcegraph/issues/23010), which will allow rebuilding SCSS files independently from each other.

I agree that speeding our frontend dev tooling is important; that's why we allocated a significant time to such improvements on our [roadmap](https://sourcegraph.productboard.com/feature-board/2330177-web-frontend-platform) and landed a couple of important ones already. After completing current goals, we will revisit modern build tools alternatives once again.
